<script>
    Function.prototype.bind2 = function(context) {
        // 调用 bind 不是函数，报错
        if(typeof this !== "function") {
            throw new Error("Function.prototype.bind - what is trying to be bound is not callable");
        }

        var self = this;
        // 获取bind2函数从第二个参数到最后一个参数
        var args = Array.prototype.slice.call(arguments,1);
        console.log(args,'args')    //  ["daisy"]

        var fNOP = function() {}
        var fBound = function() {
            // 这个时候的arguments是指bind返回的函数传入的参数
            var bindArgs = Array.prototype.slice.call(arguments);
            console.log(bindArgs,'bindArgs')    // ["18"]
            return self.apply(this instanceof fNOP ? this : context, args.concat(bindArgs));
        }

        fNOP.prototype = this.prototype
        fBound.prototype = new fNOP()
        return fBound
    }
    var foo = {
        value: 1
    }
    function bar(name, age) {
        this.habit = 'shopping';
        console.log(this.value);
        console.log(name);
        console.log(age);
    }
    bar.prototype.friend = 'kevin';
    var bindFoo = bar.bind2(foo,'daisy')
    console.log(bindFoo('18'))
</script>
